source/SVInt.cpp

```diff
@@ -1690,4 +1699,107 @@ bool exactlyEqual(const SVInt& lhs, const SVInt &rhs) {
+    }
+
+    size_t words = SVInt::getNumWords(lhs.bitWidth, false);
+    // Do word-for-word wildcard comparison, being careful that one argument
+    // might nothave unknwon flags
+    for (size_t i = 0; i < words; ++i) {
+        // bitmask of bits that actually matter in the comparison
+        uint64_t mask = ~((lhs.unknownFlag ? lhs.pVal[i + words] : 0) |
+                          (rhs.unknownFlag ? rhs.pVal[i + words] : 0));
+        // One operand can be a single word if the bitWidth is small and
+        // only one has unknowns
+        uint64_t lhsWord = lhs.isSingleWord() ? lhs.val : lhs.pVal[i];
+        uint64_t rhsWord = rhs.isSingleWord() ? rhs.val : rhs.pVal[i];
+        if ((lhsWord & mask) != (rhsWord & mask))
+            return false;
+    }
+    return true;
+}
+
+
+void copyBits(uint8_t* dest, uint16_t destBitOffset, uint8_t* src, uint16_t bitLength) {
+    // Like memcpy, but at the bit level instead of bytes
+    // i.e if destBitOffset % 8 == bitLength % 8 == 0, this is
+    // equivalent to memcpy(dest + destBitOffset / 8, src, bitLength / 8)
+
+    // Get the first byte we want to write to, and the reamining bits are a bit offset
+    dest += destBitOffset / 8;
+    destBitOffset %= 8;
+    uint16_t srcBitOffset = 0;
+    while (bitLength > 0) {
+        // Number of bits we are writing to this byte
+        uint16_t bitsToWrite = std::min<uint8_t>(bitLength, 8 - destBitOffset);
+        // get the next 8 bits of src, probably not byte aligned
+        uint8_t srcByte = (*src >> srcBitOffset) + (src[1] << (8 - srcBitOffset));
+
+        *dest = (*dest   & ((1 << destBitOffset) - 1)) + // preserved bits
+                ((srcByte & ((1 << bitsToWrite) - 1)) << destBitOffset); // new bits
+        // next write will begin at the next byte boundry
+        destBitOffset = 0;
+        bitLength -= bitsToWrite;
+        ++dest;
+        srcBitOffset += bitsToWrite;
+        src += srcBitOffset / 8;
+        srcBitOffset %= 8;
+    }
+}
+
+SVInt concatenate(ArrayRef<SVInt> operands) {
+    // First, compute how many bits total we are dealing with
+    uint16_t bits = 0;
+    bool unknownFlag = false;
+
+    for (const auto& op : operands) {
+        bits += op.bitWidth;
+        unknownFlag |= op.unknownFlag;
+    }
+
+    // words is the count of not unknown words
+    uint16_t words = SVInt::getNumWords(bits, false);
+    if (words == 1 && !unknownFlag) {
+        // The concatenation still fits into a single word
+        uint16_t offset = 0;
+        uint64_t val = 0;
+        // The first operand writs the msb, therefore, we must operate in reverse
+        for (const SVInt* op = operands.end() - 1; op >= operands.begin(); --op) {
+            copyBits((uint8_t*)&val, offset, (uint8_t*)&op->val, op->bitWidth);
+            offset += op->bitWidth;
+        }
+        return SVInt(val);
+    }
+    uint64_t *data = new uint64_t[words * (unknownFlag + 1)]();
```

Comment:
Note that the () on the end here causes the data to be zeroed, so no need for the memset.